### PR TITLE
UnityTls Integration

### PR DESCRIPTION
**Pull request by AndreasReich**
c# Purpose & Motivation

This adds the UnityTLS library as a TLS backend to curl.
UnityTLS is a new TLS wrapper library that provides unity with platform independent TLS support and unifies previously scattered use of various TLS libraries in a single interface.
By integrating UnityTLS into curl we no longer need to worry about TLS backends and dependencies in curl which will also potentially decrease binary sizes.

# Building
To enable build with `without-ssl with-unitytls` on Mac/Linux/.. or with `WITH_UNITYTLS=TRUE` on Windows.

# Relevant Implementation Details
- curl as executable does not work unless you manually enable (preprocessor) `UNITYTLS_LINKED`.  `UNITYTLS_LINKED` is for testing only and a bit hard to setup, since you need to link directly against UnityTLS which is quite cumbersome for now and will not work everywhere. Otherwise, a program using the curl library needs to call `curl_unitytls_set_interface`. This is similar to how we do the Mono integration right now.
- We use the platform specific system CA store if no CA file(s) have been provided but verification is still enabled. This deviates a bit from what some impls do, but has obvious advantages.
- We don't support right now...
  - Sessions (need extension of unitytls interface)
  - Public key pinning (could implement but is not needed right now)
  - Certificate info (https://curl.haxx.se/libcurl/c/CURLOPT_CERTINFO.html, few impls support it, unclear what we would need for that)
  - Setting SSL crypto engines (most impls don't support it)
  - Checking for pending data (few impls support it, would need extension of unitytls interface)
  - CRL (would need to extend unitytls)

# Testing
Tested modified unity version (branch waiting for this to go in) on Mac and Windows. Passes all relevant Unity tests.

Using `UNITYTLS_LINKED` and a bit of hackery with experimental dynamic library modules I was able to run Curl's own extensive tests.
**Tl;dr:** HTTP RFC1867-type formposting is broken for unknown reason, but this should be unrelated (config issue?). HTTPS tests succeed otherwise, however we fail not with the correct error in a few cases. Other than that we're in good shape!

I built/configured curl similarly as we do it for Unity on most platforms:
```-arch x86_64  -mmacosx-version-min=10.11 -lMacStandalonePlayer_TLSModule_Dynamic " CXXFLAGS="-g -arch x86_64  -mmacosx-version-min=10.11 -lMacStandalonePlayer_TLSModule_Dynamic "
LDFLAGS="-g -arch x86_64  -mmacosx-version-min=10.11 -lMacStandalonePlayer_TLSModule_Dynamic "
LIBS=-lMacStandalonePlayer_TLSModule_Dynamic 
sh configure --prefix=/Users/andreasreich/dev/curl/temp/macosx64   --disable-sftp --disable-scp --disable-ldap --disable-ldaps --disable-dict --disable-telnet --disable-tftp --disable-pop3 --disable-imap --disable-smtp --disable-gopher --disable-manual --disable-verbose --disable-sspi --disable-unix-sockets --disable-versioned-symbols --enable-static --enable-shared --without-libidn --without-libssh2 --enable-threaded-resolver --without-ssl --with-unitytls
```
(command line a bit polluted due to generating script).

Direct output:
```
TESTDONE: 631 tests out of 643 reported OK: 98%
TESTFAIL: These test cases failed: 311 312 313 506 507 580 581 1013 1014 1026 1119 1506
TESTDONE: 1047 tests were considered during 432 seconds.
TESTINFO: 404 tests were skipped due to these restraints:
TESTINFO: "curl lacks sftp server support" 33 times (582, 583, 600, 602, 604, 606, 608, 609, 610 and 24 more)
TESTINFO: "curl lacks Metalink support" 16 times (2005, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015 and 7 more)
TESTINFO: "curl lacks TrackMemory support" 2 times (96, 558)
TESTINFO: "precheck command error" 4 times (241, 1083, 1085, 2023)
TESTINFO: "rlimit problem: select limit is FD_SETSIZE 1024" 2 times (518, 537)
TESTINFO: "curl lacks NTLM support" 49 times (67, 68, 69, 70, 81, 89, 90, 91, 150 and 40 more)
TESTINFO: "curl lacks NTLM_WB support" 1 times (1310)
TESTINFO: "curl lacks PSL support" 1 times (1136)
TESTINFO: "curl lacks gopher server support" 3 times (1200, 1201, 1202)
TESTINFO: "curl lacks http/2 server support" 3 times (1700, 1701, 1702)
TESTINFO: "configured as DISABLED" 10 times (594, 836, 882, 938, 1209, 1211, 1316, 1510, 1512 and 1 more)
TESTINFO: "curl lacks scp server support" 11 times (601, 603, 605, 607, 617, 619, 621, 623, 629 and 2 more)
TESTINFO: "curl lacks telnet support" 2 times (1326, 1327)
TESTINFO: "curl lacks idn support" 5 times (165, 1034, 1035, 2046, 2047)
TESTINFO: "curl lacks OpenSSL support" 2 times (307, 308)
TESTINFO: "curl lacks TLS-SRP support" 5 times (320, 321, 322, 323, 324)
TESTINFO: "curl lacks unittest support" 22 times (1300, 1301, 1302, 1303, 1304, 1305, 1306, 1307, 1308 and 13 more)
TESTINFO: "curl lacks gopher-ipv6 server support" 1 times (1203)
TESTINFO: "curl lacks debug support" 46 times (823, 832, 835, 869, 878, 881, 907, 934, 937 and 37 more)
TESTINFO: "curl lacks unix-sockets support" 2 times (1435, 1436)
TESTINFO: "curl lacks WinSSL support" 1 times (2043)
TESTINFO: "curl lacks tftp server support" 16 times (271, 283, 284, 285, 286, 1007, 1009, 1049, 1093 and 7 more)
TESTINFO: "curl lacks crypto support" 47 times (64, 65, 72, 88, 153, 154, 167, 168, 175 and 38 more)
TESTINFO: "curl lacks sftp support" 1 times (632)
TESTINFO: "curl lacks http/2 support" 1 times (1800)
TESTINFO: "curl lacks pop3 server support" 31 times (850, 851, 852, 853, 854, 855, 856, 857, 858 and 22 more)
TESTINFO: "curl lacks SSLpinning support" 7 times (2034, 2035, 2037, 2038, 2041, 2042, 2048)
TESTINFO: "curl lacks smtp server support" 43 times (900, 901, 902, 903, 904, 908, 909, 910, 911 and 34 more)
TESTINFO: "curl lacks imap server support" 37 times (800, 801, 802, 803, 804, 805, 806, 807, 808 and 28 more)
```

My annotations to it:
```
# HTTP - all passing (or skipped)
# FTP - all passing (or skipped)
# FILE - all passing (or skipped)

# HTTPS
311     # HTTPS wrong subjectAltName but right CN -- we give out connection error when it should be verification error
312     # HTTPS GET to localhost and null-prefixed CN cert -- we give out connection error when it should be verification error
313     # CRL Test -- we don't support revoked certs

# FTPS - all passing (or skipped)

# libcurl source code tests, not using the curl command tool
506     # HTTP with shared cookie list (and dns cache) -- not passed by OpenSSL version either
507     # multi interface get with non-existing host name       -- getting verify failure when we should get "couldn't connect to host" unclear why this happens
580     # multi interface, multiple Location: headers
581     # multi interface, multiple Content-Type: headers

# SCP/SFTP - all passing (or skipped)
# SOCKS4/5 - all passing (or skipped)
# IMAP - all passing (or skipped)
# POP3 - all passing (or skipped)
# SMTP - all passing (or skipped)

# miscellaneous
1013    # Compare curl --version with curl-config --protocols   -- not dangerous and unrelated
1014    # Compare curl --version with curl-config --features
1026    # curl --manual  -- not passed by openssl version either

# unit tests - all passing (or skipped)
# libcurl source code tests, not using the curl command tool
1506    # HTTP GET connection cache limit (CURLMOPT_MAXCONNECTS) -- not passed by openssl version either

# unit tests - all passing (or skipped)
# multiple sequential protocols per test case - all passing (or skipped)
```

### CODE REVIEW

File: lib/vtls/unitytls.c [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+  unitytls_x509verify_result verifyresult = unitytls->unitytls_tlsctx_process_handshake(connssl->ctx, &err);
+  CURLcode result = CURLE_OK;
+
+  if (err.code == UNITYTLS_USER_WOULD_BLOCK) {
+    return CURLE_OK;  /* all fine but no state change yet */
+  }
+
+  if(verifyresult != UNITYTLS_X509VERIFY_SUCCESS) {
+    if(verifyresult == UNITYTLS_X509VERIFY_FATAL_ERROR)
+      failf(data, "Cert handshake failed. verify result: UNITYTLS_X509VERIFY_FATAL_ERROR. error state: %i", err.code);
```
Comment (bitter) [inline lib/vtls/unitytls.c]: This is most likely not a verification error. Maybe use a different return code and not CURLE_PEER_FAILED_VERIFICATION;

File: lib/vtls/unitytls.c [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+
+/*
+* Implementation
+*
+*/
+
+static char* load_file(const char* filepath, long* out_size)
+{
+  char* filecontent;
+  FILE* file = fopen(filepath, "rb");
```
[the remark this answers is not in the document]
Comment (AndreasReich) [reply lib/vtls/unitytls.c]: fopen is redefined by curl to `curl_fopen` - see memdebug.h
I'd hope that this one is fine

File: lib/vtls/unitytls.c [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+  WIN32_FIND_DATAA file_data;
+  char filename[MAX_PATH];
+  HANDLE hFind;
+
+  if(err->code != UNITYTLS_SUCCESS)
+    return false;
+
+  /* Path needs to end with '\*' */
+  memset(filename, 0, MAX_PATH);
+  memcpy(filename, path, len);
```
[the remark this answers is not in the document]
Comment (AndreasReich) [reply lib/vtls/unitytls.c]: good catch! fixed

File: lib/vtls/unitytls.c (hunk 1 above)
Comment (AndreasReich) [reply lib/vtls/unitytls.c]: I agree. I had this returning CURLE_SSL_CONNECT_ERROR in an earlier version, but this made the curl tests less happy. Do you think we should stick to CURLE_SSL_CONNECT_ERROR  anyways?

**Comment by AndreasReich** [comment]
I think I found the source of most of the test failures: They are caused by the missing curlssl_random implementation!
Curl fails those tests with `curl: (27) PRNG seeding failed`

File: lib/vtls/unitytls.c (hunk 1 above)
Comment (bitter) [reply lib/vtls/unitytls.c]: not sure, maybe we need to figure out when curl expect a versification error and when it doesn't. it's possible some of the fatal errors we have are actually verification errors.

File: lib/vtls/unitytls.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (bitter) [reply lib/vtls/unitytls.c]: ah, okay

File: lib/vtls/unitytls.c (hunk 1 above)
Comment (AndreasReich) [reply lib/vtls/unitytls.c]: I reconsidered and went with `CURLE_SSL_CONNECT_ERROR` for `UNITYTLS_X509VERIFY_FATAL_ERROR` now again.

This makes test "0311...[HTTPS wrong subjectAltName but right CN]" fail with "curl returned 35, when expecting 51" since unitytls does not recognize this kind of error as a verification failure. Interestingly, there is a lot of special handling code about this particular piece of verification in curl's openssl implementation while there is nothing in the corresponding mbedtls implementation. (although admittedly I haven't jumped through the necessary hoops to run those tests with mbedtls)
The way our verification code translation works for mbedtls, it seems that we cannot fix this problem for unitytls' mbed backend (we would never give out `UNITYTLS_X509VERIFY_FATAL_ERROR` for any kind of verification error).

**Comment by Wumpf** [comment]
This is much more extra work than I anticipated. I exposed unitytls random number generator since - due to the errors curl was reporting - I assumed this is the reason for the test failure in the HTTP Post tests (which I'm now determined to fix, since it *is* a serious issue after all!). Turned out that this didn't change a thing, only that curl no longer falls back to it's own random methods which is great but doesnt solve the problem at hand.
Now, I  am suspecting that I need to implement `curlssl_data_pending` which is not actually hard in either OpenSSL or Mbed, but UnityTLS does not provide it yet at all and I'm not sure yet how to approach this since I'd rather do a unitytls_tlsctx_read call and check if it succeeds.

**Comment by AndreasReich** [comment]
[deleted earlier comment, made a mistake in setup]

Exposing unitytls random generator fixes a ton of important test failures, I updated the PR desc accordingly. Change in connection error code lets fail a bit more often with generic sll connection error when it should be a  auth error according to Curl.

In my opinion the only serious test failures remaining are:

- 301     # HTTPS GET with user and password
- 313     # CRL Test

which are both because of features we have decided not to support.

**Comment by bitter** [comment]
Why is "301 # HTTPS GET with user and password" failing?

**Comment by bitter** [comment]
Would be nice to know why #301 and #306 fails before pushing this to master

**Comment by AndreasReich** [comment]
Acknowledged! Will dig deeper into it. Currently I only know that both fail with receive errors

**Comment by AndreasReich** [comment]
Found the source of both test failures and was able to fix it: If someone would gracefully close a connection, the next read caused a UNITYTLS_INTERNAL_ERROR. I fixed this in UnityTls and now we're passing 301 and 306 as well! :)

(updated PR desc again)
